crypto/bsslmlkem/mlkem768.c

```diff
@@ -0,0 +1,1140 @@
+                        const uint8_t *message,
+                        const uint8_t *randomness)
+{
+    int du = kDU768;
+    int dv = kDV768;
+    uint8_t counter = 0;
+    vector secret, error;
+    uint8_t input[33];
+    scalar scalar_error;
+    vector u;
+    scalar v;
+    scalar expanded_message;
+
+    vector_generate_secret_eta_2(&secret, &counter, randomness);
+    vector_ntt(&secret);
+    vector_generate_secret_eta_2(&error, &counter, randomness);
+    memcpy(input, randomness, 32);
+    input[32] = counter;
+    scalar_centered_binomial_distribution_eta_2_with_prf(&scalar_error, input);
+    matrix_mult(&u, &pub->m, &secret);
+    vector_inverse_ntt(&u);
+    vector_add(&u, &error);
+    scalar_inner_product(&v, &pub->t, &secret);
+    scalar_inverse_ntt(&v);
+    scalar_add(&v, &scalar_error);
+    scalar_decode_1(&expanded_message, message);
+    scalar_decompress(&expanded_message, 1);
+    scalar_add(&v, &expanded_message);
+    vector_compress(&u, du);
+    vector_encode(out, &u, du);
+    scalar_compress(&v, dv);
+    scalar_encode(out + compressed_vector_size(RANK768), &v, dv);
+}
+
+/*
+ * See section 6.2.
+ * entropy[MLKEM_ENCAP_ENTROPY])
+ */
+static void mlkem_encap_external_entropy(uint8_t *out_ciphertext,
+                                         uint8_t *out_shared_secret,
+                                         const public_key_RANK768 *pub,
+                                         const uint8_t *entropy)
+{
+    uint8_t input[64];
+    uint8_t key_and_randomness[64];
+
+    memcpy(input, entropy, MLKEM_ENCAP_ENTROPY);
+    memcpy(input + MLKEM_ENCAP_ENTROPY, pub->public_key_hash,
+           sizeof(input) - MLKEM_ENCAP_ENTROPY);
+    hash_g(key_and_randomness, input, sizeof(input));
+    encrypt_cpa(out_ciphertext, pub, entropy, key_and_randomness + 32);
+    memcpy(out_shared_secret, key_and_randomness, 32);
+}
+
+/*
+ * out_ciphertext[ossl_mlkem768_CIPHERTEXT_BYTES],
+ * out_shared_secret[ossl_mlkem768_SHARED_SECRET_BYTES],
+ * entropy[MLKEM_ENCAP_ENTROPY])
+ */
+static
+void ossl_mlkem768_encap_external_entropy(uint8_t *out_ciphertext,
+                                          uint8_t *out_shared_secret,
+                                          const ossl_mlkem768_public_key *public_key,
+                                          const uint8_t *entropy)
+{
+    const struct public_key_RANK768 *pub =
+        public_key_768_from_external(public_key);
+
+    mlkem_encap_external_entropy(out_ciphertext, out_shared_secret, pub, entropy);
+}
+
+/* Calls |ossl_mlkem768_encap_external_entropy| with random bytes from |RAND_bytes| */
+int ossl_mlkem768_encap(uint8_t *out_ciphertext,
+                        uint8_t *out_shared_secret,
+                        const ossl_mlkem768_public_key *public_key)
+{
+    uint8_t entropy[MLKEM_ENCAP_ENTROPY];
+
+    if (!mlkem_init())
+        return 0;
+
+    RAND_bytes(entropy, MLKEM_ENCAP_ENTROPY);
```

Comment:
> RAND_bytes_ex - and check the return value.

So changed in https://github.com/openssl/openssl/pull/25848/commits/467e70f9af6925357f8f418f4e47e354c0a9fa15

> Also we might want a variant of this function which takes an external entropy input (instead of generating it here) in order for us to do KAT testing.

As he's working on this, I'd defer to the contribution by @andrewkdinh to add such function.